### PR TITLE
[Bugfix](sampling): search Qrita top-p pivots in log space

### PR DESCRIPTION
**Pull request by qianlihuang**
## Purpose

Search Qrita top-p pivots in log-probability space to improve accuracy on peaked distributions without regressing performance.

## Test Plan

Run an inline Python benchmark in the vLLM 0.25.0 container that generates and JIT-compiles both the before and after Triton kernels, then compares them against the PyTorch reference and benchmarks the `384 × 154,880` workload:

<details>

```bash
python3 - <<'PY'
import ast
import gc
import importlib.util
import statistics
import sys
import tempfile
from pathlib import Path

import torch
import vllm
import vllm.v1.sample.ops.topk_topp_triton as installed
from vllm.v1.sample.ops.topk_topp_sampler import apply_top_k_top_p_pytorch


class QritaVariant(ast.NodeTransformer):
    """Generate the before or after version of PR #48927."""

    def __init__(self, after: bool):
        self.after = after
        self.in_kernel = False
        self.pivots = 0
        self.exits = 0

    @staticmethod
    def expr(source):
        return ast.parse(source, mode="eval").body

    @staticmethod
    def assigns_name(statement, name):
        return (
            isinstance(statement, ast.Assign)
            and len(statement.targets) == 1
            and isinstance(statement.targets[0], ast.Name)
            and statement.targets[0].id == name
        )

    def visit_FunctionDef(self, node):
        old = self.in_kernel
        if node.name == "_topk_topp_kernel":
            self.in_kernel = True
        self.generic_visit(node)
        self.in_kernel = old
        return node

    def visit_Assign(self, node):
        self.generic_visit(node)

        if (
            self.in_kernel
            and len(node.targets) == 1
            and isinstance(node.targets[0], ast.Name)
            and node.targets[0].id == "p_pivot_0"
        ):
            node.value = self.expr(
                "tl.sqrt(max_range * tl.maximum(min_range, 1e-30))"
                if self.after
                else "(max_range - min_range) * 0.5 + min_range"
            )
            self.pivots += 1

        return node

    def visit_If(self, node):
        self.generic_visit(node)

        if not self.in_kernel or "num_iters >= 18" not in ast.unparse(node.test):
            return node

        # Only the three top-p fallback branches assign p_pivot directly.
        pivot_assignment = next(
            (
                statement
                for statement in node.body
                if self.assigns_name(statement, "p_pivot")
            ),
            None,
        )
        if pivot_assignment is None:
            return node

        node.test = self.expr(
            (
                "((max_range - min_range) < 1e-9 "
                "and max_range <= min_range * 1.0001) "
                "or num_iters >= 18"
            )
            if self.after
            else "(max_range - min_range) < 1e-9 or num_iters >= 18"
        )

        pivot_assignment.value = self.expr(
            "p_pivot_0"
            if self.after
            else "(max_range + min_range) / 2.0"
        )
        self.exits += 1
        return node


def load_variant(source, module_name, after, directory):
    tree = ast.parse(source)
    transform = QritaVariant(after)
    tree = transform.visit(tree)
    ast.fix_missing_locations(tree)

    if (transform.pivots, transform.exits) != (3, 3):
        raise RuntimeError(
            "Unexpected Qrita structure: "
            f"pivots={transform.pivots}, exits={transform.exits}"
        )

    path = Path(directory) / f"{module_name}.py"
    path.write_text(ast.unparse(tree))

    spec = importlib.util.spec_from_file_location(module_name, path)
    module = importlib.util.module_from_spec(spec)
    sys.modules[module_name] = module
    spec.loader.exec_module(module)
    return module


def make_peaked_logits(batch, vocab, seed):
    generator = torch.Generator(device="cuda").manual_seed(seed)
    logits = torch.randn(
        batch,
        vocab,
        device="cuda",
        dtype=torch.float32,
        generator=generator,
    )
    hot = torch.randint(
        vocab,
        (batch, 8),
        device="cuda",
        generator=generator,
    )
    boosts = torch.tensor(
        [13.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0],
        device="cuda",
    ).expand_as(hot)
    logits.scatter_add_(1, hot, boosts)
    return logits


@torch.inference_mode()
def accuracy_case(name, logits, k, p, variants):
    reference = apply_top_k_top_p_pytorch(logits.clone(), k, p)
    reference_probs = reference.softmax(dim=-1)

    print(name)
    for variant_name, function in variants:
        output = function(logits.clone(), k, p)
        tv = 0.5 * (
            reference_probs - output.softmax(dim=-1)
        ).abs().sum(dim=-1)

        print(
            f"  {variant_name:6s} "
            f"TV mean={tv.mean().item():.6e}, "
            f"max={tv.max().item():.6e}"
        )

    del reference, reference_probs


@torch.inference_mode()
def latency_case(name, logits, k, p, variants):
    print(name)

    for variant_name, function in variants:
        # Compile and warm up.
        for _ in range(5):
            output = function(logits.clone(), k, p)
        torch.cuda.synchronize()

        samples = []
        for _ in range(20):
            start = torch.cuda.Event(enable_timing=True)
            end = torch.cuda.Event(enable_timing=True)

            start.record()
            output = function(logits.clone(), k, p)
            end.record()
            end.synchronize()

            samples.append(start.elapsed_time(end))

        print(
            f"  {variant_name:6s} "
            f"median={statistics.median(samples):.3f} ms, "
            f"mean={statistics.mean(samples):.3f} ms"
        )


if not torch.cuda.is_available():
    raise RuntimeError("CUDA GPU required")

source_path = Path(installed.__file__)
source = source_path.read_text()

with tempfile.TemporaryDirectory() as directory:
    before_module = load_variant(
        source, "_qrita_before_48927", False, directory
    )
    after_module = load_variant(
        source, "_qrita_after_48927", True, directory
    )

    variants = [
        ("before", before_module.apply_top_k_top_p_triton),
        ("after", after_module.apply_top_k_top_p_triton),
    ]

    vocab = 154_880

    print(f"vLLM: {vllm.__version__}")
    print(f"GPU:   {torch.cuda.get_device_name(0)}")
    print(f"File:  {source_path}")
    print()

    # Fast correctness check: only four rows.
    accuracy_logits = make_peaked_logits(4, vocab, 42)
    accuracy_p = torch.full(
        (4,), 0.95, device="cuda", dtype=torch.float32
    )
    accuracy_k = torch.full(
        (4,), 32_000, device="cuda", dtype=torch.int32
    )

    accuracy_case(
        "Accuracy: standalone top-p",
        accuracy_logits,
        None,
        accuracy_p,
        variants,
    )
    accuracy_case(
        "Accuracy: fused top-k + top-p",
        accuracy_logits,
        accuracy_k,
        accuracy_p,
        variants,
    )

    del accuracy_logits, accuracy_p, accuracy_k
    gc.collect()
    torch.cuda.empty_cache()
    print()

    # PR workload: 384 × 154,880. No slow PyTorch reference here.
    perf_logits = make_peaked_logits(384, vocab, 43)
    perf_p = torch.full(
        (384,), 0.95, device="cuda", dtype=torch.float32
    )
    perf_k = torch.full(
        (384,), 32_000, device="cuda", dtype=torch.int32
    )

    latency_case(
        "Latency: standalone top-p, 384 x 154880",
        perf_logits,
        None,
        perf_p,
        variants,
    )
    latency_case(
        "Latency: fused top-k + top-p, 384 x 154880",
        perf_logits,
        perf_k,
        perf_p,
        variants,
    )
PY
```

</details>

Environment:

```text
vLLM: 0.25.0
GPU: NVIDIA GeForce RTX 5090
```
Workload simulates GLM-5.2 (top_p=0.95) serving at 1024 decode concurrency with DEP8 and MTP3, yielding 384 sampling rows per rank over a 154,880-token vocabulary.

## Test Result

```text
vLLM: 0.25.0
GPU:   50Series
File:  /usr/local/lib/python3.12/dist-packages/vllm/v1/sample/ops/topk_topp_triton.py

Accuracy: standalone top-p
  before TV mean=1.510686e-02, max=2.288285e-02
  after  TV mean=4.230319e-06, max=8.308022e-06
Accuracy: fused top-k + top-p
  before TV mean=1.163947e-02, max=2.188168e-02
  after  TV mean=1.368576e-05, max=2.996346e-05

Latency: standalone top-p, 384 x 154880
  before median=3.012 ms, mean=3.017 ms
  after  median=2.850 ms, mean=2.847 ms
Latency: fused top-k + top-p, 384 x 154880
  before median=1.958 ms, mean=1.958 ms
  after  median=1.967 ms, mean=1.967 ms
```

---

<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

* [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
* [x] The test plan, such as providing test command.
* [x] The test results, such as pasting the results comparison before and after, or e2e results
* [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

</details>

### CODE REVIEW

**Comment by qianlihuang** [comment]
I’d prefer to keep this PR as a draft for now. The synthetic workload exposes a real issue, but I haven’t yet validated its impact on real model distributions, dataset-level quality, or acceptance-rate shifts. And I’m not yet sure that the log-space pivot is the best fix, especially since this changes the Qrita algorithm.

Also, I opened #48928 to optimize the common MTP top-p-only path used by many default generation configs, so this PR may not need to be rushed.
